### PR TITLE
[Inductor] Add support for NEON ISA in the Inductor C++ backend

### LINKED ISSUE #104729: Add support for NEON ISA in the Inductor C++ backend
### 🚀 The feature, motivation and pitch

Context: The TorchInductor C++ backend currently supports vectorization in C++ Codegen through two Intel ISAs: AVX2 and AVX512, as mentioned in the [Update 5 Blog](https://dev-discuss.pytorch.org/t/torchinductor-update-5-cpu-backend-backend-performance-update-and-deep-dive-on-key-optimizations/1117#vectorization-in-c-codegen-4). While the Aten library does support Arm as well, we are yet to leverage its NEON/SVE ISAs to generate optimized kernels. The blog also [mentions](https://dev-discuss.pytorch.org/t/torchinductor-update-5-cpu-backend-backend-performance-update-and-deep-dive-on-key-optimizations/1117#vectorization-in-c-codegen-4:~:text=It%20can%20be,sub%2Dclasses.) that the VecISA class can be subclassed in order to support other ISAs.

Proposal: I am working on providing NEON ISA support for the TorchInductor's C++ backend. Particularly, I intend to provide a NEON implementation of the `vec_reduce_all()` function, which currently has optimized [AVX2 and AVX512 intrinsics implementations](https://github.com/pytorch/pytorch/blob/ced5c89b6fbe827a538b7ada96b2f9a5989871c7/aten/src/ATen/cpu/vec/functional_base.h#L37-L79) for x86 processors introduced by @mingfeima in #73953, as well as a [slow path](https://github.com/pytorch/pytorch/blob/ced5c89b6fbe827a538b7ada96b2f9a5989871c7/aten/src/ATen/cpu/vec/functional_base.h#L12-L28) implementation for other processors including Arm. I have implemented a NEON version for the function, wired up the Inductor's generated C++ to invoke this NEON path on Arm CPUs & I've seen performance improvements, particularly in the Softmax operation.

Posting this here for any discussion before raising a PR.

### Alternatives

_No response_

### Additional context

_No response_

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### PR DESCRIPTION
**Pull request by Rohanjames1997**
Fixes #104729

As suggested in the [blog](https://dev-discuss.pytorch.org/t/torchinductor-update-5-cpu-backend-backend-performance-update-and-deep-dive-on-key-optimizations/1117#:~:text=It%20can%20be,sub%2Dclasses.), I subclassed the `VecISA` class and implemented a NEON version of the `vec_reduce_all()` function, to go along with the existing AVX2 and AVX512 versions. Any operation that calls `vec_reduce_all()` will also take the NEON path and benefit from its vectorization.

The `vec_reduce_all()` is invoked by Softmax and other operations like norms. Using the fast path results in 30% time savings for Softmax as compared to the previously taken slow path.

  | Slow path | Fast path (NEON intrinsics)
-- | -- | --
Softmax (100 passes, 1024 dimension) | 623.706ms | 452.011ms

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang @Xia-Weiwen @ngimel @malfet

### CODE REVIEW

File: aten/src/ATen/cpu/vec/functional_base.h [hunk 1]
```diff
@@ -76,6 +76,34 @@ struct VecReduceAllSIMD<float, Op> {
...
+    // 32-bit shuffle: [a1+a3+a5+a7, a2+a4+a6+a8, a1+a3+a5+a7, a2+a4+a6+a8, -, -, -, -] -> [a2+a4+a6+a8, a1+a3+a5+a7, a2+a4+a6+a8, a1+a3+a5+a7, -, -, -, -]
+    v1_1 = vrev64q_f32(v.get_low());
+    v1 = {v1_1, v1_1};
+    // [a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, -, -, -, -]
+    v = vec_fun(v, v1);
+
+    return v.get_low()[0];
+  }
+};
+#endif // defined(CPU_CAPABILITY_NEON)
```
Comment (Rohanjames1997) [inline aten/src/ATen/cpu/vec/functional_base.h:106]: Reference AVX-2 implementation: [here](https://github.com/pytorch/pytorch/blob/ced5c89b6fbe827a538b7ada96b2f9a5989871c7/aten/src/ATen/cpu/vec/functional_base.h#L38-L56)

Comment (jgong5) [reply aten/src/ATen/cpu/vec/functional_base.h:106]: Do we have UT to cover this?

**Review by jgong5 — COMMENTED** [review]
Overall LGTM, will stamp after we make sure the changes are covered by UT and the PR passes the UT.

File: aten/src/ATen/cpu/vec/functional_base.h (hunk 1 above)
Comment (Rohanjames1997) [reply aten/src/ATen/cpu/vec/functional_base.h:106]: @jgong5 , I need some help in writing the UT for this.

One option is to test if the accuracy of reduce_all or softmax is comparable in eager and compiled mode. Let me know if this is acceptable.

Other option is to extend Intel's AVX `vec_reduce_all` UT. From what I understood from this [comment](https://github.com/pytorch/pytorch/pull/73953#issuecomment-1121820121), the UT for Intel's `vec_reduce_all` is accomplished by `vec_test_all_types_AVX2`, `vec_test_all_types_AVX512` and `vec_test_all_types_DEFAULT`. From the build logs, I see that these test executables are built by cmake. My guess is that if I need to extend the existing UT structure and generate a similar executable for NEON, I would need to modify certain **critical build files** such as this and more cc @malfet https://github.com/pytorch/pytorch/blob/795885d9476a4978c476b221720332b73d994d7d/cmake/Codegen.cmake#L315-L350 
Although my current code passes the [FunctionalTestsReducedFloat test](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L1429), it goes through the default path. Could you share some more details on how I should go about adding a UT for this reduction function? 

Thanks!

Comment (Rohanjames1997) [reply aten/src/ATen/cpu/vec/functional_base.h:106]: @jgong5 , I need some help in writing the UT for this.

One option is to test if the accuracy of `reduce_all` or `softmax` is comparable in eager and compiled mode. Let me know if this is acceptable.

Other option is to extend Intel's AVX `vec_reduce_all` UT. From what I understood from this [comment](https://github.com/pytorch/pytorch/pull/73953#issuecomment-1121820121), the UT for Intel's `vec_reduce_all` is accomplished by `vec_test_all_types_AVX2`, `vec_test_all_types_AVX512` and `vec_test_all_types_DEFAULT`. From the build logs, I see that these test executables are built by cmake. My guess is that if I need to extend the existing UT structure and generate a similar executable for NEON, I would need to modify certain **critical build files** such as this and more. https://github.com/pytorch/pytorch/blob/795885d9476a4978c476b221720332b73d994d7d/cmake/Codegen.cmake#L315-L350
Although my current code passes the [FunctionalTestsReducedFloat test](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L1429), it goes through the default path. Could you share some more details on how I should go about adding a UT for this reduction function?

Thanks!

Comment (Rohanjames1997) [reply aten/src/ATen/cpu/vec/functional_base.h:106]: Additionally, do the CI UTs run on x86 machines? Could you throw some light on how NEON tests are executed? cc @malfet

**Comment by Rohanjames1997** [comment]
Thanks for the review @jgong5
I decided to extend the existing UT infrastructure for NEON as well. Since my changes involve editing some critical cmake files as well, I've raised the changes in a [different PR](https://github.com/pytorch/pytorch/pull/105763/files/b2cbedfa28d6d66e62ce53ab29dee33aa52a511d..58fbfebe5d7317ed83a0475fe01bca8b3e2391b7) and my code passes the UT `vec_test_all_types_NEON`. 

Let me know if you'd like me to merge those changes into this PR, or keep them separate.
Thanks!

**Comment by jgong5** [comment]
> Let me know if you'd like me to merge those changes into this PR, or keep them separate.

 How about make them separate and let the other PR land first?

**Comment by Rohanjames1997** [comment]
@jgong5 , Sure I have raised the changes exclusively required to create a UT in this PR: #105823

**Comment by Rohanjames1997** [comment]
@jgong5 @malfet, Could you help remove the stale label and add a no-stale label to this PR?
Additionally, could the PR #105823 be reviewed? I realize the PT2.1 release process was ongoing, but I'd be glad to do any further testing to get the two PRs merged. Thanks!

**Comment by Rohanjames1997** [comment]
@jgong5, can you please reopen this PR and #105823 ? I plan to resolve conflicts and continue further work on it. Thanks!

**Comment by Rohanjames1997** [comment]
I've fixed the merge conflicts again and this PR is ready for review along with #105823.

Hi @kimishpatel @digantdesai @jerryzh168,
Could you review these PRs?

**Review by jgong5 — DISMISSED** [review]

**Review by malfet — DISMISSED** [review]

**Comment by Rohanjames1997** [comment]
To witness the perf gains of this PR, follow these simple steps:

1. Spawn an aarch64 instance (I tested mostly on Graviton3 instances).
2. Checkout PyTorch to the main branch & build it.
3. Run the script attached below as follows: `OMP_NUM_THREADS=1 LRU_CACHE_CAPACITY=1024 DNNL_DEFAULT_FPMATH_MODE=BF16 python TestSoftmax.py`. Note that I am using a single thread for simplicity, but perf gains will be witnessed at higher thread counts too.
4. Checkout PyTorch to my branch & build it. (Rohanjames1997:neon_vec_reduce_all) 
5. Run Step 3 

You will see that the compiled mode performance improves by around 30% as stated in the PR description. 

To test accuracy, I have used `torch.allclose()` between the eager mode and the compiled mode outputs. 

**To test accuracy using the existing tests in aten/src/ATen/test/vec_test_all_types.cpp, let's look at #105823** 

Here are the contents of TestSoftmax.py are printed below for easy reference 
````
import torch
import torch.nn as nn
from torch.profiler import profile, record_function, ProfilerActivity

model = nn.Softmax().eval()
compiled_model = torch.compile(model)
inputs = torch.randn(1024, 1024)

with torch.set_grad_enabled(False):
    for _ in range(50):
        compiled_model(inputs) #Warmup
    print("Warmup over")
    with profile(activities=[ProfilerActivity.CPU]) as prof:
        with record_function("model_inference"):
            for _ in range(100):
                compiled_model(inputs)

print(prof.key_averages().table(sort_by="self_cpu_time_total"))
# Check if the compiled model inference and the eager model inference are similar using torch.allclose
print(torch.allclose(compiled_model(inputs), model(inputs)))
````


[TestSoftmax.txt](https://github.com/pytorch/pytorch/files/14370884/TestSoftmax.txt)

**Comment by kit1980** [comment]
> @pytorchmergebot merge -i

@Rohanjames1997 please be careful with ignoring, you ignored a lint failure.

File: torch/_inductor/codegen/cpp_prefix.h [hunk 2]
```diff
@@ -19,7 +19,7 @@
 #include <c10/util/Half.h>
 #include <c10/util/TypeCast.h>
 
-#if defined(CPU_CAPABILITY_AVX512) || defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_ZVECTOR)
+#if defined(CPU_CAPABILITY_AVX512) || defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_ZVECTOR) || defined(CPU_CAPABILITY_NEON)
```
Comment (kit1980) [inline torch/_inductor/codegen/cpp_prefix.h:22]: Line too long.

Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:22]: For all `# elif defined(CPU_CAPABILITY_ZVECTOR)` in the `masked_load` functions in this file (e.g., https://github.com/pytorch/pytorch/pull/105590/files#diff-e384e0d2829ef483854a45c4f422979d1a2ca28495c7bc06e91eabc67c61a470R320), change them to `# else` so that the default path can work for NEON too.

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by Rohanjames1997** [comment]
Refer to #123584
